pkg/server/endpoints/endpoints_test.go

```diff
@@ -3,335 +3,765 @@ package endpoints
+	})
+	require.NoError(t, err)
+	assert.Equal(t, tcpAddr, endpoints.TCPAddr)
+	assert.Equal(t, udsAddr, endpoints.UDSAddr)
+	assert.Equal(t, svidObserver, endpoints.SVIDObserver)
+	assert.Equal(t, testTD, endpoints.TrustDomain)
+	assert.NotNil(t, endpoints.RegistrationServer)
+	assert.NotNil(t, endpoints.NodeServer)
+	if assert.NotNil(t, endpoints.ExperimentalServers) {
+		assert.NotNil(t, endpoints.ExperimentalServers.AgentServer)
+		assert.NotNil(t, endpoints.ExperimentalServers.BundleServer)
+		assert.NotNil(t, endpoints.ExperimentalServers.EntryServer)
+		assert.NotNil(t, endpoints.ExperimentalServers.SVIDServer)
+	}
+	assert.NotNil(t, endpoints.BundleServer)
+	assert.Equal(t, ds, endpoints.DataStore)
+	assert.Equal(t, log, endpoints.Log)
+	assert.Equal(t, metrics, endpoints.Metrics)
 }
 
-func (s *EndpointsTestSuite) SetupTest() {
-	s.ds = fakedatastore.New(s.T())
+func TestListenAndServe(t *testing.T) {
+	ca := testca.New(t, testTD)
+	serverSVID := ca.CreateX509SVID(serverID)
+	agentSVID := ca.CreateX509SVID(agentID)
+	adminSVID := ca.CreateX509SVID(adminID)
+	downstreamSVID := ca.CreateX509SVID(downstreamID)
+
+	dir, err := ioutil.TempDir("", "")
+	require.NoError(t, err)
+	defer func() {
+		os.RemoveAll(dir)
+	}()
+
+	listener, err := net.Listen("tcp", "localhost:0")
+	require.NoError(t, err)
+	require.NoError(t, listener.Close())
 
+	udsPath := filepath.Join(dir, "socket")
+
+	ds := fakedatastore.New(t)
 	log, _ := test.NewNullLogger()
-	ip := net.ParseIP("127.0.0.1")
-	td := url.URL{
-		Scheme: "spiffe",
-		Host:   "example.org",
+	metrics := fakemetrics.New()
+
+	registrationServer := newRegistrationServer()
+	nodeServer := newNodeServer()
+	bundleServer := newBundleServer()
+
+	endpoints := Endpoints{
+		TCPAddr:            listener.Addr().(*net.TCPAddr),
+		UDSAddr:            &net.UnixAddr{Name: udsPath, Net: "unix"},
+		SVIDObserver:       newSVIDObserver(serverSVID),
+		TrustDomain:        testTD,
+		DataStore:          ds,
+		RegistrationServer: registrationServer,
+		NodeServer:         nodeServer,
+		ExperimentalServers: &ExperimentalServers{
+			AgentServer:  &agentv1.UnimplementedAgentServer{},
+			BundleServer: &bundlev1.UnimplementedBundleServer{},
+			EntryServer:  &entryv1.UnimplementedEntryServer{},
+			SVIDServer:   &svidv1.UnimplementedSVIDServer{},
+		},
+		BundleServer: bundleServer,
+		Log:          log,
+		Metrics:      metrics,
 	}
 
-	s.mockClock = clock.NewMock(s.T())
-	s.mockClock.Set(time.Now())
-
-	catalog := fakeservercatalog.New()
-	catalog.SetDataStore(s.ds)
-
-	s.svidState = svid.State{}
-	c := &Config{
-		TCPAddr: &net.TCPAddr{IP: ip, Port: 8000},
-		UDSAddr: &net.UnixAddr{Name: "/tmp/spire-registration.sock", Net: "unix"},
-		SVIDObserver: svid.ObserverFunc(func() svid.State {
-			return s.svidState
-		}),
-		TrustDomain: td,
-		Catalog:     catalog,
-		Log:         log,
+	// Prime the datastore with the:
+	// 1- bundle used to verify client certificates.
+	// 4- agent attested node information
```

Comment:
nit: I'm not sure what the numbers stand for. Is the order intentional?

Reply:
Those are stale... I'll remove the numbering.